docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,237 @@
+# Proposal: Standardizing the Alloy standard library
+
+* Author: Robert Fratto (@rfratto)
+* Last updated: 2024-06-27
+* Discussion link: <https://github.com/grafana/alloy/pull/1161>
+
+## Abstract
+
+This proposal introduces a way to standardise on how new capabilities are added
+to Alloy's standard library so that usage of the standard library feels
+consistent.
+
+## Problem
+
+As of Alloy v1.2.0, there are 19 identifiers in the standard library:
+
+* `base64_decode`
+* `coalesce`
+* `concat`
+* `constants`
+* `env`
+* `format`
+* `join`
+* `json_decode`
+* `json_path`
+* `nonsensitive`
+* `replace`
+* `split`
+* `to_lower`
+* `to_upper`
+* `trim`
+* `trim_prefix`
+* `trim_space`
+* `trim_suffix`
+* `yaml_decode`
+
+These identifiers were added organically over time; there has been little
+thought to consistency across all identifiers.
+
+While 19 identifiers is manageable and possible for a user to remember the most
+important ones, if we don't establish a way to standardise existing and new
+identifiers, the inconsistency between identifiers will only grow over time as
+new identifiers are added and become harder to understand.
+
+This proposal aims to:
+
+* Establish a convention for how to name identifiers in the standard library.
+* Propose a way to align existing identifiers with the new convention without
+  breaking backwards compatibility.
+
+For the sake of demonstrating the convention, new identifiers are provided as
+examples, but are out of scope of this proposal and should not be considered
+accepted if this proposal is accepted.
+
+## Proposal
+
+The standard library will introduce namespaces for identifiers that group
+similar identifiers together:
+
+* Array functions
+    * `array.concat` (previously `concat`)
+* Conversion functions
+    * `convert.nonsensitive` (previously `nonsensitive`)
+* Decoding functions
+    * `decode.base64` (previously `decode_base64`)
+    * `decode.json` (previously `decode_json`)
+    * `decode.yaml` (previously `decode_yaml`)
+* OS-related functions
+    * `os.env` (previously `env`)
```

Comment:
I'm a little worried `os` is a bad namespace since I can't identify any new functionality that would belong in it. The examples I gave later (`os.uid`, `os.gids`) are contrived without an obvious use case. 

However, one possibility is we learn that `os` is a more appropriate name for what is currently called `constants`. Then we could have: 

* `os.env` (previously `env`) 
* `os.name` (previously `constants.os`) 
* `os.hostname` (previously `constants.hostname`) 
* `os.arch` (previously `constants.arch`) 

I believe this would be the most effective use of the `os` namespace (to fill it with constants and utilities). I would like to treat this as out of scope for this proposal as much as we can, but it is still important for us to make sure that the namespace names I've selected make sense.

Reply:
I don't think `os` is terrible, but how about `system`?
- `system.env` - env variables
- `system.name` = `constants.os`
- `system.hostname` = `constants.hostname`
- system.arch` = `constants.arch`

Reply:
`system` seems like a good candidate. How can we evaluate `system` and `os`? The most obvious difference between the two is character length, but I don't think that should be the only criteria we look at.

Reply:
IMO `os` is more specific and `system` is more general. Operating System means the lowest-level software on the hardware like Linux and Windows. System can encompass hardware and software. And in this example, the `system.arch` I think works better, as it typically will match the hardware. 

So I think `system` or `sys` is a better option overall. Not all the things we add into this namespace may be related to OS - for example `system.cpu_count` or `system.memory`. It's also a namespace often used in other language standard libraries, like Java, while Golang uses os mostly.

Reply:
That makes sense to me :+1: I do prefer `sys` just to keep it as short as possible, and I think `sys` is a common enough abbreviation that it won't be too confusing. 

If you think sys will work, I'll update the proposal

Reply:
Yep, I like sys, it's what [Python](https://docs.python.org/3/library/sys.html) uses too.